### DISCUSSION TITLE
An "assets.transform" hook to allow the addition of new asset format transformations

### DISCUSSION THREAD

**Discussion by ChappIO** [directus/directus#17362] [Feature Requests]
This feature request originates from this question: https://github.com/directus/directus/discussions/17349

**Some Background**

I am building a website to share my music. As I don't feel like spending my time building a backoffice instead of making music, Directus would almost be the perfect solution.

The only feature I am still lacking is being able to generate low-quality previews of my music from the high-quality (downloadable) .wav files to save my visitor's bandwidth. I currently convert these in my backend for frontend (NextJS) but it is a hassle to manage. In fact, I have already converted the image processing to Directus because it works a lot better than NextJS.

**The Feature**

I would like to be able to plug in my own asset format transformation so I can extend the `GET /asset` endpoint that is already in Directus.

I think one solution would be by allowing users of Directus to register a filter hook called something like `assets.transform` where we can implement transformers for more mime types other than the hardcoded image ones.

This would then be called from `AssetsService.ts#getAsset` if the requested transformation was not found in the storage.

In my case I wouldn't even require any changes to the endpoint itself as the `format=mp3&quality=5` parameters would do fine (ffmpeg takes a quality index between 0 and 9).

**Things I've Tried: Pull Request**

I initially started by trying to implement this myself but I couldn't get the development environment up and running.
My issues:

 - When I try to run `pnpm install` it asks me which version on `@directus/shared` to install (which says it is for internal purposes).
 - When I try to run the e2e tests (even though my laptop is chugging hard with all the docker compose databases running) I run into oracledb installation issues.

I would be very happy to help develop this feature as I am using it for my own project anyway.

**Things I've Tried: Custom Endpoint**

Now this version would work. And if I can't get any support getting my dev environment up and running, I will go this route (hell I may even do it anyway to get my website up sooner). 

This endpoint works by:

1. Fetch the file from directus by id
2. Check its mime type (must be audio/x-wav)
3. Calculate the asset suffix just as `AssetsService.ts#getAssetSuffix` does
4. Check if the file exists in the `StorageManager`, if it does: return it, if it does not: generate it

While this works, I do believe providing this functionality as a hook (or maybe even in-core for auto) would be very valuable to more people than just myself.

<sub>sorry for the wall of text... I tried to keep it concise</sub>

**Comment by ChappIO** [comment]
If I could get pointers to how to set up my development environment, I would still be willing to contribute a PR.

**Reply by paescuj** [reply]
Have you already been on this page? https://docs.directus.io/contributing/running-locally.html

Please note that the feature request must have been accepted first before we can look into any pull requests. You'll find more information on that here: https://docs.directus.io/contributing/introduction.html

**Reply by ChappIO** [reply]
I did but I missed a detail there. It works now. Thanks.

**Comment by ChappIO** [comment]
For context, I hacked together an extension which provides this functionality in a custom endpoint: https://gist.github.com/ChappIO/12fda5ad3b1bbc9ca612ed3967d8146e

You can call it using: `GET /audio/<file-id>/<download-filename>?quality=<0-9>&download`

It requires `ffmpeg` to be installed to run the `fluent-mpeg` package.

Though I believe allowing devs to plug in asset transformers would be a huge boon.

**Reply by ChappIO** [reply]
Also, if it were integrated as a hook I could drop 90% of the logic in the extension, since most of it is:

1. Should I generate the file?
2. How to serve the file with range support?
3. What should the suffix hash be?

All things that are already handled by `AssetsController` and `AssetsService`.

**Comment by rijkvanzanten** [comment]
Couple of additional questions to flesh this out out of the top of my head:

- Do we start to accept any query param in the assets endpoint to pass along to the extension? Eg format/quality are "native", but you might want to have a custom `bitrate`? 🤔 
  - How does that affect caching?
- What would be the IO of the transformer hook? Would it be a stream input and stream output with a metadata object with the file information?
- Should you be able to overwrite the regular image one somehow? 
  - If not, how should we handle the scenario where Directus itself might include mp3 support over time?

Excited about this! Adding this as an extension point really opens up the "do whatever" attitude that we're using in the rest of the platform as well 🙂

**Reply by ChappIO** [reply]
> Do we start to accept any query param in the assets endpoint to pass along to the extension? Eg format/quality are "native", but you might want to have a custom bitrate? 🤔
> How does that affect caching?

2 Options I can see:

I think a good starting point would be to not allow additional parameters. This would make it simpler and not alter the public API. Instead, I think if extension developers want to add specific parameters, they should do that through the "key" parameters.

For example: I could catch a "key=preview" parameter and generate an asset using whatever implementation I deem appropriate for audio previews.

An other option would be to expose a more complicated "AssetsTransformerService" interface instead of hooks. This would allow a step where the transformer can declare which parameters it is interested in. Allowing the AssetsService still deal with the (suffix based) caching based on those used parameters.

> What would be the IO of the transformer hook? Would it be a stream input and stream output with a metadata object with the file information?

I think that depends on the method chosen. If we go for a hook without additional parameters, I think the IO would be:

 Input:
  - The (source) File Item: I might want to check for mime types and such
  - A function to open a ReadableStream to the original asset
  
 Output:
  - A new ReadableStream or undefined
  
 If a ReadableStream was returned, the data will be piped to the cache and the ReadableStream will be closed by the AssetsService:
 
 ```javascript
 filter('asset.transform', ({file, options, openRead, openWrite}) => {
  if (file.type === 'audio/x-wav' && (options.format === 'mp3' || options.filename_download?.endsWith('.mp3')) {
     // I can transform this!
     return ffmpeg(openRead())
       .fromFormat('wav')
       .toFormat('mp3')
       .quality(5);
  }
});
 ```
 
Though for future backward compatibility, an object containing metadata might be more useful. Then any (optional) parameter could be added later:

 ```javascript
 filter('asset.transform', ({file, options, openRead, openWrite}) => {
  if (file.type === 'audio/x-wav' && (options.format === 'mp3' || options.filename_download?.endsWith('.mp3')) {
     // I can transform this!
     const stream = ffmpeg(openRead())
       .fromFormat('wav')
       .toFormat('mp3')
       .quality(5));
       
     return {stream};
  }
});
 ```
 
 In either case AssetsService could just have:
 
 ```javascript
 
 transformAsset(file, transformation)) {
    const suffix = calculateSuffix(transformation);
    if(!hasCache(file, suffix)) {
       const result = launchFilterHook('assets.transform', {whateverInput});
       if(result) {
         result.stream.pipe(openWriteStream(), {end: true});
         await new Promise((resolve) => {
            result.on('end', () => resolve();
            
            // deal with errors here
         });
       }
    }
    
    return whatever the rest of the service needs.
 }
 ```


> Should you be able to overwrite the regular image one somehow?
> If not, how should we handle the scenario where Directus itself might include mp3 support over time?~

I think it should be possible to override yes. If Directus would decide to add support for a media type, and that would break an existing implementation, that would be a bummer.

Also: I like the "do whatever" attitude. This would also fit that thought.

**Reply by rijkvanzanten** [reply]
I actually really like the idea of making the individual 'mime type handlers' it's own service type override 🤔 That fits neatly in with how we're handling file storage abstraction, and some of those other vendor driver specific things. That way, we could ship with a couple default ones (just the one for images for now), and people can insert their own as an extension type 🤔 Like you mentioned, it could have a `.register()` function that would return an object declaring what mimetypes it'll handle, and what parameters it expects, which the API can then use in the regular caching :o

**Reply by paescuj** [reply]
Me too! 👍 What would happen if the same type is registered multiple times? Maybe we would like to support some kind of pipeline as well 🤔

**Reply by rijkvanzanten** [reply]
> Me too! 👍 What would happen if the same type is registered multiple times? Maybe we would like to support some kind of pipeline as well 🤔

I think in that case the one that's registered last will take control. That way, you could override the system behavior for modifying the defaults 🤔

**Reply by paescuj** [reply]
> I think in that case the one that's registered last will take control. That way, you could override the system behavior for modifying the defaults 🤔

That would make sense! 👍

**Comment by 100100101** [comment]
I ran into a similar situation.
I need the assets images to the client to get round in png format.
I read Sharp documentation and realized that making images round with [Operation, ...arguments] won't work. Well, or this solution will not be beautiful.
What @ChappIO suggested seems
```JavaScript
filter('asset.transform', (...) => {
    //
})
```
perfectly logical to me.
What is currently the best/correct way to implement rounding photos, and in general to implement such scenarios?

@rijkvanzanten thanks for this amazing product.

**Reply by ChappIO** [reply]
> I ran into a similar situation.
> 
> I need the assets images to the client to get round in png format.
> 
> I read Sharp documentation and realized that making images round with [Operation, ...arguments] won't work. Well, or this solution will not be beautiful.
> 
> What @ChappIO suggested seems
> 
> ```JavaScript
> 
> filter('asset.transform', (...) => {
> 
>     //
> 
> })
> 
> ```
> 
> perfectly logical to me.
> 
> What is currently the best/correct way to implement rounding photos, and in general to implement such scenarios?
> 
> 
> 
> @rijkvanzanten thanks for this amazing product.


What I ended up doing was creating a custom endpoint with caching myself. This way I had a bit more control over the processing parameters.

**Reply by 100100101** [reply]
Did it end up looking like a 
```ts
import type { Application } from 'express'
export default ({ init }) => {
    init('middlewares.after', ({ app }: Record<'app', Application>) => {
        app.use((_req, res, next) => {
            const { baseUrl } = _req
            // Or a similar condition
            if (baseUrl !== '/assets') return next()
            console.log('Here is the data change', 'Modify response')
            // ...
            next()
        })
    })
}
```

?

**Reply by 100100101** [reply]
I get it, you said endpoint not hook
Could you show what it looked like?

**Comment by 100100101** [comment]
@rijkvanzanten 
Essentially I need to implement something like [https://github.com/lovell/sharp#stream](https://github.com/lovell/sharp#stream), but input is a Buffer, so this preset didn't work:
```json
[
    ...
    [
        "composite",
        [
            {
                "input": "<svg><rect x=\"0\" y=\"0\" width=\"50\" height=\"50\" rx=\"10\" ry=\"10\"/></svg>",
                "blend": "dest-in"
            }
        ]
    ],
    ...
]
```
If there is no option to make a preset via json, then I will do it via monkey patching `res.end res.write` or `AssetsService`

**Comment by 100100101** [comment]
@rijkvanzanten
Finally made a hook
```ts
import { HookConfig } from '@directus/shared/types'
const hook: HookConfig = (register, { services: { AssetsService } }) => {
    const assetsServiceGetAssetOriginal = AssetsService.prototype.getAsset
    AssetsService.prototype.getAsset = async function (this: any, ...args) {
        const { transforms } = args[1]
        if (transforms?.length) {
            for (const transform of transforms) {
                const [transformName] = transform
                if (transformName !== 'composite') continue
                const compositeRulesOptions = transform[1]
                if (!Array.isArray(compositeRulesOptions)) continue
                for (const compositeRuleOptions of compositeRulesOptions) {
                    const compositeOptionsInput = compositeRuleOptions?.input
                    if (typeof compositeOptionsInput !== 'string') continue
                    compositeRuleOptions.input = Buffer.from(
                        compositeOptionsInput
                    )
                }
            }
        }
        return assetsServiceGetAssetOriginal.call(this, ...args)
    }
}
export default hook
```
Now you can use like this

![Снимок экрана 2023-04-20 в 02 00 45](https://user-images.githubusercontent.com/12983864/233224350-fac96639-dd02-4512-b72f-e688d4fb9bb9.png)


<img width="1346" alt="Снимок экрана 2023-04-20 в 01 29 53" src="https://user-images.githubusercontent.com/12983864/233221564-e6ceda5a-1eeb-4aa6-8794-9bd5affd0909.png">

And also I want to take this opportunity to say that I made a utility for loading ts from extensions/*/index.js . This allows you to use compilation on the fly. Later I want to use this approach with esm.
I also import node inspector
But I'll write more about that later in another thread.

<img width="1209" alt="Снимок экрана 2023-04-20 в 01 48 13" src="https://user-images.githubusercontent.com/12983864/233223819-aa4db4f7-cada-4159-a4c8-4ad83c431451.png">

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
